docs/P4-Compilation.md

```diff
@@ -0,0 +1,131 @@
+Compiling your own P4 code
+==========================
+
+This document explains how to compile your own P4 code, that can be located anywhere outside the open-P4studio tree.
+
+## A little theory...
+
+On the surface, the P4 compiler for Tofino (`bf-p4c`) works like any other compiler. In other words, it is totally ok to compile your program using the following command line:
+
+```
+bf-p4c my_program.p4
+```
+
+This should work for any simple program, written for Tofino. At the end of the compilation process the compiler will create the directory `my_program.tofino`, containing all compilation artifacts. If your program is intended for Tofino2, add the parameter `--target tofino2`; the name of the compilation directory will be `my_program.tofino2` then.
+
+In more complex cases, you might need to provide addiional parameters, similar to the ones used by most C compilers, such as `-I <include_dir>`, `-DCPP_VAR[=value]` or `-UCPP_VAR`. It is also highly recommended to add the `-g` parameter, so that the compiler can output the detailed logs and placement information (especially if you have access to the `p4i` (P4Insight) tool). 
+
+The compiler has many other special options and they might sometimse be required, but they are not the focus of this document.
+
+While the command above is the easiest way to invoke the compiler and is often used by the developers to check whether the program compiles or not, it is rarely used. The reason is that after compiling a program we need to exercise it by starting the user-space driver and loading the program onto the target, be it either a Tofino model or the real ASIC. In the simulation case it is also important to allow the model to load the program layout information so that it can produce human-readable logs by translating low-level operations back into the familiar symbols from your program. 
+
+Generally speaking, the compiler produces three main artifacts that are neccessary to run a P4 program:
+
+1. The compiled program binary, called `tofino.bin` (or `tofino2.bin`) that needs to be loaded into the ASIC or the model so that they can execute the P4 code
```

Comment:
When I built open-p4studio, and added a symbolic link named `bf-p4c` that pointed to `p4c`, I could run the `bf-p4c` command above, but it did not create any files with a suffix of `.bin`.  It did produce a file with a suffix of `.bfa`, which I think contained Tofino assembly.

Should I have linked `bf-p4c` to something else instead of `p4c`?  Or could open-p4studio have problems, if it is not creating `.bin` files without additional command line options?

Reply:
This sounds pretty strange and should not be happening. Let me try to reproduce. Are you sure `$SDE_INSTALL/bin/bfas` is there and can be executed? I also assume you are talking about the file(s) inside `my_program.tofino/pipe` directory.

Reply:
I can confirm seeing the same thing on my system.  

I can see that the compiler driver has been significantly modified, compared to Intel's SDE and specifically a new parameter, `--enable-bf-asm` has been introduced. By default it is not set, which is why `bfas` is not invoked.

@fruffy -- any specific reasons for these changes?

Reply:
bf-asm is not supplied with the compiler here: https://github.com/p4lang/p4c/tree/main/backends/tofino

Instead it is available in the studio here: 
https://github.com/p4lang/open-p4studio/tree/main/pkgsrc/p4-compilers

The option is required to be able to build the compiler in both environments. I do not know why bf-asm is split this way.

Reply:
Thanks, @fruffy! I suspect that this might've happened because Intel didn't want to release the register map initially, although given this project that decision makes very little sense. 

I'll file an issue, because as I explained, one must be running the assembler every time they compile the program even if they do not think they might need the binary. 

One way to fix it is to utilize the internal check inside the compiler driver that distinguishes between the "DEVELOPER" and "INSTALLED" environments. Another is to have an option '--disable-bf-asm' instead.

The bottom line is that once `open-p4studio` is installed, the compiler driver must (by default) run the assembler afterwards.

Reply:
We could merge this back into the p4c repository. That would simplify code a bunch and make the compiler more usable out of the box. 

But I am currently working on this on my spare time, so not sure when I can get around to it.